gpiozero/pins/data.py

```diff
@@ -1078,6 +1078,101 @@ def pulled_up(self, function):
+            return spec
+        else:
+            if isinstance(spec, bytes):
+                spec = spec.decode('ascii')
+            spec = spec.upper()
+            if spec.isdigit():
+                return self.to_gpio(int(spec))
+            if spec.startswith('GPIO') and spec[4:].isdigit():
+                return self.to_gpio(int(spec[4:]))
+            elif spec.startswith('BCM') and spec[3:].isdigit():
+                return self.to_gpio(int(spec[3:]))
+            elif spec.startswith('WPI') and spec[3:].isdigit():
+                main_head = 'P1' if 'P1' in self.headers else 'J8'
+                try:
+                    return self.to_gpio({
+                        0:  '%s:11' % main_head,
+                        1:  '%s:12' % main_head,
+                        2:  '%s:13' % main_head,
+                        3:  '%s:15' % main_head,
+                        4:  '%s:16' % main_head,
+                        5:  '%s:18' % main_head,
+                        6:  '%s:22' % main_head,
+                        7:  '%s:7'  % main_head,
+                        8:  '%s:3'  % main_head,
+                        9:  '%s:5'  % main_head,
+                        10: '%s:24' % main_head,
+                        11: '%s:26' % main_head,
+                        12: '%s:19' % main_head,
+                        13: '%s:21' % main_head,
+                        14: '%s:23' % main_head,
+                        15: '%s:8'  % main_head,
+                        16: '%s:10' % main_head,
+                        17: 'P5:3',
+                        18: 'P5:4',
+                        19: 'P5:5',
+                        20: 'P5:6',
+                        21: '%s:29' % main_head,
+                        22: '%s:31' % main_head,
+                        23: '%s:33' % main_head,
+                        24: '%s:35' % main_head,
+                        25: '%s:37' % main_head,
+                        26: '%s:32' % main_head,
+                        27: '%s:36' % main_head,
+                        28: '%s:38' % main_head,
+                        29: '%s:40' % main_head,
+                        30: '%s:27' % main_head,
+                        31: '%s:28' % main_head,
+                        }[int(spec[3:])])
+                except KeyError:
+                    raise PinInvalidPin('%s is not a valid wiringPi pin' % spec)
+            elif ':' in spec:
+                header, pin = spec.split(':', 1)
+                if pin.isdigit():
+                    try:
+                        header = self.headers[header]
+                    except KeyError:
+                        raise PinInvalidPin(
+                            'there is no %s header on this Pi' % header)
+                    try:
+                        function = header.pins[int(pin)].function
+                    except KeyError:
+                        raise PinInvalidPin(
+                            'no such pin %s on header %s' % (pin, header.name))
```

Comment:
Took me a while to figure out why the PinInvalidPin exception on line 1160 used `header` but the PinInvalidPin exception on line 1166 used `header.name` - I _eventually_ realised it's because line 1158 does `header = self.headers[header]` :confused:  (i.e. the type of the `header` variable depends on whether that line threw an exception or not! :nauseated_face: )